server/routes/users.ts

```diff
@@ -16,46 +18,94 @@ router.get('/signOut', async (req, res) => {
 
 // POST /signIn
 router.post('/signIn', async (req, res) => {
-    if(req.session.user)
+    if(req.session.user) // already logged in
         return res.status(403).json({ message: 'Unauthorized. User already logged in.' });
 
-    if(!req.headers.authorization)
+    if(!req.headers.authorization)// no auth token provided
         return res.status(403).json({ message: 'Unauthorized. Need token for login.' });
 
-    try {
+    let uid: string, email: string;
+    try { // authenticate with firebase
         let idToken: string = req.headers.authorization as string;
-        idToken = idToken.split('JWT ')[1];
-        const { email, uid } = await getAuth().verifyIdToken(idToken);
-        // TODO: also check if email and uid match database values??? Might be unnecesary
-        const user = { _id: uid, email: email };
-        req.session.user = user;
-        return res.json({ message: 'Logged in.' });
+        if(!idToken || !checkNonEmptyString(idToken))
+            throw new Error("[routes/users POST signIn] id token is invalid.");
+        idToken = idToken.split('JWT ')[1];  
+        const decodedToken: DecodedIdToken = await getAuth().verifyIdToken(idToken);
+        uid = decodedToken.uid;
+        email = decodedToken.email;
     } catch(e) {
         return res.status(403).json({ message: 'Unauthorized. Failed to verify token.', error: e.message });
     }
+
+    try { // ensure user is in the database
+        await getUserByEmail(email);
+    } catch(e) {
+        return res.status(401).json({ message: 'User does not exist in database, try sign up.', error: e.message });
+    }
+
+    // establish express session for current user
+    const user = { _id: uid, email: email };
+    req.session.user = user;
+    return res.json({ message: 'Signed in.' });
```

Comment:
Should redirect to main dash

Reply:
Are we handling redirection on backend? I assumed we would leave it up to front end to handle to maintain the SPA invariant of our website.

I see our backend as simply a way to get data.